File: 3583-count-special-triplets/3583-count-special-triplets.cpp
#define MOD 1000000007
class Solution {
public:
    int specialTriplets(vector<int>& nums) {
        int res = 0;
        unordered_map<int, int> freqs;
        unordered_map<int, int> leftFreqs;
        for(int i = 0; i < nums.size(); i++)
        {
            freqs[nums[i]]++;
        }
        for(int i = 0; i < nums.size(); i++)
        {
            long long leftCount = leftFreqs.find(nums[i] * 2) != leftFreqs.end() ? leftFreqs[nums[i] * 2] : 0;
            leftFreqs[nums[i]]++;
            long long rightCount = freqs[nums[i] * 2] - leftFreqs[nums[i] * 2];
            res = (res + ((rightCount * leftCount) % MOD)) % MOD;
        }
        return res;
    }
};

File: 3583-count-special-triplets/3583-count-special-triplets.py
class Solution:
    def specialTriplets(self, nums: List[int]) -> int:
        res = 0
        MOD = 1000000007
        freqs = Counter(nums)
        leftFreqs = {}
        for num in nums:
            leftCount = leftFreqs[num * 2] if num * 2 in leftFreqs else 0
            leftFreqs[num] = leftFreqs.get(num, 0) + 1
            rightCount = freqs[num * 2] - leftCount if num > 0 else freqs[0] - leftFreqs[0]
            res = (res + ((rightCount * leftCount) % MOD)) % MOD;
        return res